File: go_engine/engine/workflow.go
package engine

import "fmt"

type Command struct {
	CommandType string
	Attributes  map[string]interface{}
}

func ScheduleActivity(commandID, activityID, activityName string, input map[string]interface{}, timeoutSeconds int, heartbeatTimeoutSeconds int, retryPolicy RetryPolicy) Command {
	return Command{
		CommandType: "ScheduleActivity",
		Attributes: map[string]interface{}{
			"command_id":              commandID,
			"activity_id":             activityID,
			"activity_name":           activityName,
			"input":                   input,
			"timeout_seconds":         timeoutSeconds,
			"heartbeat_timeout_seconds": heartbeatTimeoutSeconds,
			"retry_policy":            retryPolicy,
		},
	}
}

func ScheduleTimer(commandID, timerID string, timeoutSeconds int) Command {
	return Command{
		CommandType: "ScheduleTimer",
		Attributes: map[string]interface{}{
			"command_id":     commandID,
			"timer_id":       timerID,
			"timeout_seconds": timeoutSeconds,
		},
	}
}

func CompleteWorkflow(result map[string]interface{}) Command {
	return Command{
		CommandType: "CompleteWorkflow",
		Attributes:  map[string]interface{}{"result": result},
	}
}

func FailWorkflow(error string) Command {
	return Command{
		CommandType: "FailWorkflow",
		Attributes:  map[string]interface{}{"error": error},
	}
}

type DecisionResult struct {
	NextState string
	StateData map[string]interface{}
	Commands  []Command
	Complete  bool
	Failure   string
}

type WorkflowDefinition interface {
	Name() string
	InitialState() string
	Decide(ctx *WorkflowContext, state string, stateData map[string]interface{}, history []HistoryEvent) DecisionResult
}

type WorkflowContext struct {
	RunID     string
	StateData map[string]interface{}
	History   []HistoryEvent
	sequence  int
}

func NewWorkflowContext(runID string, stateData map[string]interface{}, history []HistoryEvent) *WorkflowContext {
	seq := 0
	if val, ok := stateData["sequence"].(float64); ok {
		seq = int(val)
	}
	if val, ok := stateData["sequence"].(int); ok {
		seq = val
	}
	return &WorkflowContext{
		RunID:     runID,
		StateData: stateData,
		History:   history,
		sequence:  seq,
	}
}

func (c *WorkflowContext) NextSequence() int {
	c.sequence++
	c.StateData["sequence"] = c.sequence
	return c.sequence
}

func (c *WorkflowContext) EnsureCommandID(prefix, key string) string {
	if existing, ok := c.StateData[key].(string); ok && existing != "" {
		return existing
	}
	cmdID := fmt.Sprintf("%s-%d", prefix, c.NextSequence())
	c.StateData[key] = cmdID
	return cmdID
}

func (c *WorkflowContext) WorkflowInput() map[string]interface{} {
	for _, event := range c.History {
		if event.EventType == "WorkflowStarted" {
			if input, ok := event.Attributes["input"].(map[string]interface{}); ok {
				return input
			}
		}
	}
	return map[string]interface{}{}
}

func (c *WorkflowContext) CurrentTime() float64 {
	var latest float64
	for _, event := range c.History {
		if event.EventType == "WorkflowTaskStarted" || event.EventType == "WorkflowStarted" {
			latest = event.Timestamp
		}
	}
	return latest
}

func (c *WorkflowContext) ActivityCompleted(activityID string) bool {
	for _, event := range c.History {
		if event.EventType == "ActivityCompleted" && event.Attributes["activity_id"] == activityID {
			return true
		}
	}
	return false
}

func (c *WorkflowContext) ActivityResult(activityID string) map[string]interface{} {
	for _, event := range c.History {
		if event.EventType == "ActivityCompleted" && event.Attributes["activity_id"] == activityID {
			if result, ok := event.Attributes["result"].(map[string]interface{}); ok {
				return result
			}
		}
	}
	return nil
}

func (c *WorkflowContext) TimerFired(timerID string) bool {
	for _, event := range c.History {
		if event.EventType == "TimerFired" && event.Attributes["timer_id"] == timerID {
			return true
		}
	}
	return false
}

File: python_engine/workflow_engine/workflow.py
"""
Workflow definition interfaces and deterministic context.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, Iterable, List, Optional, Protocol

from .models import HistoryEvent, RetryPolicy


@dataclass
class Command:
    command_type: str
    attributes: Dict[str, Any]


@dataclass
class ScheduleActivity(Command):
    def __init__(
        self,
        command_id: str,
        activity_id: str,
        activity_name: str,
        input: Dict[str, Any],
        timeout_seconds: int,
        heartbeat_timeout_seconds: int,
        retry_policy: RetryPolicy,
    ) -> None:
        super().__init__(
            command_type="ScheduleActivity",
            attributes={
                "command_id": command_id,
                "activity_id": activity_id,
                "activity_name": activity_name,
                "input": input,
                "timeout_seconds": timeout_seconds,
                "heartbeat_timeout_seconds": heartbeat_timeout_seconds,
                "retry_policy": retry_policy.__dict__,
            },
        )


@dataclass
class ScheduleTimer(Command):
    def __init__(self, command_id: str, timer_id: str, timeout_seconds: int) -> None:
        super().__init__(
            command_type="ScheduleTimer",
            attributes={"command_id": command_id, "timer_id": timer_id, "timeout_seconds": timeout_seconds},
        )


@dataclass
class CompleteWorkflow(Command):
    def __init__(self, result: Dict[str, Any]) -> None:
        super().__init__(
            command_type="CompleteWorkflow",
            attributes={"result": result},
        )


@dataclass
class FailWorkflow(Command):
    def __init__(self, error: str) -> None:
        super().__init__(
            command_type="FailWorkflow",
            attributes={"error": error},
        )


@dataclass
class DecisionResult:
    next_state: str
    state_data: Dict[str, Any]
    commands: List[Command] = field(default_factory=list)
    complete: bool = False
    failure: Optional[str] = None


class WorkflowDefinition(Protocol):
    name: str

    def initial_state(self) -> str:
        ...

    def decide(
        self,
        ctx: "WorkflowContext",
        state: str,
        state_data: Dict[str, Any],
        history: List[HistoryEvent],
    ) -> DecisionResult:
        ...


class WorkflowContext:
    def __init__(self, run_id: str, state_data: Dict[str, Any], history: Iterable[HistoryEvent]) -> None:
        self.run_id = run_id
        self.state_data = state_data
        self._history = list(history)
        self._sequence = int(state_data.get("sequence", 0))

    def _events(self, event_type: str) -> List[HistoryEvent]:
        return [event for event in self._history if event.event_type == event_type]

    def workflow_input(self) -> Dict[str, Any]:
        for event in self._events("WorkflowStarted"):
            return event.attributes.get("input", {})
        return {}

    def current_time(self) -> float:
        latest = None
        for event in self._history:
            if event.event_type in ("WorkflowTaskStarted", "WorkflowStarted"):
                latest = event.timestamp
        return latest if latest is not None else 0.0

    def next_sequence(self) -> int:
        self._sequence += 1
        self.state_data["sequence"] = self._sequence
        return self._sequence

    def ensure_command_id(self, prefix: str, key: str) -> str:
        stored = self.state_data.get(key)
        if stored:
            return stored
        command_id = f"{prefix}-{self.next_sequence()}"
        self.state_data[key] = command_id
        return command_id

    def activity_completed(self, activity_id: str) -> bool:
        return any(
            event.event_type == "ActivityCompleted" and event.attributes.get("activity_id") == activity_id
            for event in self._history
        )

    def activity_failed(self, activity_id: str) -> Optional[str]:
        for event in self._history:
            if event.event_type == "ActivityFailed" and event.attributes.get("activity_id") == activity_id:
                return event.attributes.get("error")
        return None

    def activity_result(self, activity_id: str) -> Optional[Dict[str, Any]]:
        for event in self._history:
            if event.event_type == "ActivityCompleted" and event.attributes.get("activity_id") == activity_id:
                return event.attributes.get("result")
        return None

    def timer_fired(self, timer_id: str) -> bool:
        return any(
            event.event_type == "TimerFired" and event.attributes.get("timer_id") == timer_id
            for event in self._history
        )

    def has_command_event(self, command_type: str, command_id: str) -> bool:
        for event in self._history:
            if event.event_type == command_type and event.attributes.get("command_id") == command_id:
                return True
        return False
